File: basic_dsa_two_hundred_sixty_three.cpp
/**
 * Basic DSA - Running Sum of 1d Array (LeetCode 1480)
 * Return running sum: nums[i] = sum(nums[0]..nums[i]).
 */
#include <iostream>
#include <vector>

std::vector<int> runningSum(std::vector<int>& nums) {
    for (int i = 1; i < (int)nums.size(); i++)
        nums[i] += nums[i - 1];
    return nums;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Running Sum of 1d Array\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {1, 2, 3, 4};
    runningSum(v1);
    std::cout << "[1,2,3,4] -> [";
    for (int i = 0; i < (int)v1.size(); i++) std::cout << (i ? "," : "") << v1[i];
    std::cout << "]\n";

    std::vector<int> v2 = {1, 1, 1, 1, 1};
    runningSum(v2);
    std::cout << "[1,1,1,1,1] -> [";
    for (int i = 0; i < (int)v2.size(); i++) std::cout << (i ? "," : "") << v2[i];
    std::cout << "]\n";
}

File: basic_dsa_two_hundred_sixty_three.py
"""
Basic DSA - Basic Calculator II
Given a string s representing a valid expression, evaluate it and return the result.
The expression contains non-negative integers and operators +, -, *, / with spaces.
Division truncates toward zero.
"""


def calculate(s):
    """
    One pass using a stack for * and / precedence.
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    stack = []
    num = 0
    op = "+"
    for i, ch in enumerate(s):
        if ch.isdigit():
            num = num * 10 + (ord(ch) - ord("0"))
        if (not ch.isdigit() and ch != " ") or i == len(s) - 1:
            if op == "+":
                stack.append(num)
            elif op == "-":
                stack.append(-num)
            elif op == "*":
                stack.append(stack.pop() * num)
            else:  # "/"
                stack.append(int(stack.pop() / num))
            op = ch
            num = 0
    return sum(stack)


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Basic Calculator II")
    print("=" * 60)

    print(calculate("3+2*2"))  # 7
    print(calculate(" 3/2 "))  # 1
    print(calculate(" 3+5 / 2 "))  # 5
